kernel: Timeslice enhancements

peter-mitsis opened the pull request:
Fixes a couple of minor issues with respect to time slicing and allows for a configurable time slice reset.

1. Calls to k_thread_time_slice_reset() used to always reset the time slice on the current CPU regardless of whether the target thread was running on the current CPU or if it was even running. We now target the correct CPU, and it only takes immediate effect if the thread that CPU's current thread.
2. Refactored thread_is_sliceable() to ensure that thread-level slice-ability does not override conditions such as the thread being prevented from executing.
3. Adds CONFIG_TIMESLICE_AUTO_RESET. If disabled, then we do not always reset the timeslice. That is, if the outgoing thread switches out before the time slice expires, then the incoming thread may continue to use the same time slice (with some restrictions) without resetting the timer. 

Update:
Dropped the auto-reset  portion as it would have introduced too much unpredictability.

peter-mitsis committed:
kernel: Inline remove_timeout()

The routine remove_timeout() is both short and only used in one
file. Inlining it helps reduce overhead during context switches
when time slicing is enabled.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: k_thread_time_slice_set to use correct CPU

When using k_thread_time_slice_set(), only reset the time slice for
the CPU on which that thread is running--if it is running.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Update thread_is_sliceable()

This fixes several minor items related to the priority or importance
of checks in determining whether the thread can be time sliced.

A thread that is prevented from running can not be time sliced
regardless of whether it was configured for thread thread-grained
time slicing or not. Nor can the idle thread be time sliced.

If the thread is configured for thread-grained time slicing, then
do not bother with the preemptible or priority threshhold checks.
This maintains the same behavior, and just optimizes the checks.

If the thread is sliceable, we may as well return the size of the
tick slice since we are checking that information anyway. Thus, a
return value of zero (0) means that the thread is not sliceable,
and a value greater than zero (0) means that it is sliceable.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Remove superfluous thread_is_sliceable() call

Within z_sched_ipi() there is no need for the thread_is_sliceable()
test as z_time_slice() performs that check. Since as a result of this
thread_is_sliceable() is now only used within timeslicing.c, the
'static' keyword is applied to it.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
test: Fix variable name

Updates the parameter name to more accurately relfect the
units of measurement (msec instead of sec).

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

andyross commented on doc/kernel/services/scheduling/index.rst:
```diff
@@ -194,6 +194,12 @@ only when dealing with lower priority threads that are less time-sensitive.
+that do not wish to incur the overhead associated with this may choose to
+disable :kconfig:option:`CONFIG_TIMESLICE_AUTO_RESET`. Disabling this may
+result with the incoming thread having a shorter than expected time slice as
+the outgoing thread already consumed a portion of it.
```
What's the use case of this feature?  It seems complicated to explain to me.  Like, apps very likely won't be using timeslicing if they have complicated IPC performance characteristics.  You use timeslicing when you have a bunch of loosely coupled threads that **don't** wait on each other.   This tunable has the smell of a footgun to me, like apps will turn it on thinking it's "faster" and then suddenly their Very Important Thread gets starved because of the some weird usage pattern.

It also seems to invert the API promise: threads might reasonably be written to assume that they get a fair share of the timeslice budget.  Now they might get less, through no fault of their own, just because someone used up "their" timeslice before switching!

Or explaining from the other side: an existing complaint of the Zephyr timeslice implementation is that it makes no effort at "large scale fairness".  You get your timeslice, but that's it, and if you give it up or whatever that's on you.  This new scheme is actually **less** fair as your timeslice size depends on the behavior of other threads.

peter-mitsis replied:
You raise a number of valid concerns that I have as well.

Right now, I see it ad trade offs. And honestly, maybe they are not worth it. It amounts to fixed size time slices for slice-able threads thereby lessening the need to futz with both the timer queue and timer hardware as much  (both of which can be slow). This gain comes with a loss--it is harder to predict in advance how long a given thread might have before it is forced to yield to another thread (if another of equal  priority exists).

If you don't think that it is worth it, I will happily drop this item--as I said, I've had similar concerns.

andyross reviewed (commented):
Notes and nitpicks.  And I'm not entirely sold on the refactoring of is_sliceable() and that auto reset feature.

peter-mitsis committed:
kernel: Encapsulate timeslice data

Creates a private timeslice data structure to contain the timeslice
data for each CPU.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis commented:
Dropped the auto-reset portion of the work.

peter-mitsis commented on kernel/include/ksched.h:
```diff
@@ -64,7 +64,7 @@ bool z_thread_prio_set(struct k_thread *thread, int prio);
 
 void z_time_slice(void);
-void z_reset_time_slice(struct k_thread *curr);
+void z_reset_time_slice(unsigned int cpu, struct k_thread *curr);
```
I left this in with today's re-push (May 15). With the existing spinlocks and abort sequence, I have not been able to spot issues. However, if you see something that I don't, please share. :)

andyross reviewed (commented):
I'll concede the aesthetics on most things.  Still think that the changes to reset are trying to fix the wrong thing in the wrong place.

peter-mitsis commented:
It has taken a while, but I think that I am starting to finally see what you mean.
